File: lib/jarvis/parser/rule.rb
module Jarvis::Parser

  class Rule
    def initialize expr, action
      @expr = expr
      @action = action
    end

    def match message
      @expr =~ message
    end

    def execute message
      matches = @expr.match message
      @action.call(matches) || true
    end
  end

end

File: lib/jarvis/parser/rule.js
export class Rule {
  constructor(expr, action) {
    this._expr = expr;
    this._action = action;
  }

  match(message) {
    return this._expr.test(message);
  }

  execute(message) {
    const matches = this._expr.exec(message);
    const result = this._action(matches);
    return result === undefined ? true : result;
  }
}

export default Rule;
